### PR TITLE
CSE-570: access token refresh, log stream reconnect

### PR DESCRIPTION
**Pull request by ekaterina-nbl**

### CODE REVIEW

File: lib/Log.js [hunk 1]
```diff
@@ -209,9 +211,11 @@ class Log {
                 return this._helper.logStream(
                     this._deviceIds, 
                     this._messageHandler.bind(this),
```
Comment (ppetrosh) [inline lib/Log.js]: BWT do we really need `.bind(this)` here? I believe the JS functions are bound to the current scope by default. Aren't they?

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "imp-central-impt",
-  "version": "1.0.4",
+  "version": "1.0.5",
```
Comment (ppetrosh) [inline package.json]: Please increase the middle version.

File: lib/util/ImpCentralApiHelper.js [hunk 3]
```diff
@@ -454,6 +463,11 @@ class ImpCentralApiHelper {
         return error instanceof ImpCentralApi.Errors.ImpCentralApiError &&
             error._statusCode === 429;
     }
+
+    _isAuthenticationError(error) {
+        return error instanceof ImpCentralApi.Errors.ImpCentralApiError &&
+            error._statusCode === 401;
```
Comment (ppetrosh) [inline lib/util/ImpCentralApiHelper.js]: or 403?

File: lib/Log.js [hunk 4]
```diff
@@ -209,9 +211,11 @@ class Log {
...
-                    this._stateChangeHandler.bind(this)).
+                    this._stateChangeHandler.bind(this),
+                    this._errorHandler.bind(this)).
                     then((logStreamId) => {
                         this._logStreamId = logStreamId;
+                        this._reconnecting = false;
```
Comment (ppetrosh) [inline lib/Log.js]: We only toggle the flag on success. What if we fail to reconnect? Are we just exiting in this case?

File: lib/util/ImpCentralApiHelper.js [hunk 5]
```diff
@@ -220,6 +204,26 @@ class ImpCentralApiHelper {
         });
     }
 
+    _refreshAccessToken() {
+        if (!this._authConfig.refreshToken && !this._authConfig.loginKey) {
+            return Promise.reject(new Errors.AccessTokenExpiredError(this._authConfig));
```
Comment (ppetrosh) [inline lib/util/ImpCentralApiHelper.js]: Is in this case access token really expired or is it a misconfiguration or something?

File: lib/util/ImpCentralApiHelper.js [hunk 6]
```diff
@@ -220,6 +204,26 @@ class ImpCentralApiHelper {
...
+                .then(result => {
+                    this._authConfig.setAccessToken(result.access_token, result.expires_at);
+                    this._impCentralApi.auth.accessToken = this._authConfig.accessToken;
+                    return this._authConfig.save();
+                }).catch(error => {
+                    if (error instanceof Errors.EntityNotFoundError ||
+                        this._isEntityNotFoundError(error)) {
+                        return Promise.reject(new Errors.RefreshTokenError(this._authConfig));
+                    }
+                    return Promise.reject(error);
```
Comment (ppetrosh) [inline lib/util/ImpCentralApiHelper.js]: Does it mean we just exit impt with an error message?

Comment (ekaterina-nbl) [reply lib/util/ImpCentralApiHelper.js]: Yes, if renew access token operation was failed, impt exit with an error

File: lib/util/ImpCentralApiHelper.js (hunk 5 above)
Comment (ekaterina-nbl) [reply lib/util/ImpCentralApiHelper.js]: It is for --temp option of impt auth login.
In this case impt does not save information (refreshToken or loginKey) required to refresh access token in .impt.auth file and there is no possibility to obtain new access token after the previous one has expired.
I added a comment to the source.

File: lib/Log.js (hunk 4 above)
Comment (ekaterina-nbl) [reply lib/Log.js]: Yes, if we fail to reconnect we are just exit through _reconnect method error handler.

File: lib/util/ImpCentralApiHelper.js (hunk 3 above)
Comment (ekaterina-nbl) [reply lib/util/ImpCentralApiHelper.js]: According to the imp central spec and my experiments, 403 is used for insufficient permission error (eg. you can see a Devicegroup but are not permitted to modify it), not for invalid authorization.
It has no sense to refresh access token for it.

File: package.json (hunk 2 above)
Comment (ekaterina-nbl) [reply package.json]: Done

File: lib/Log.js (hunk 1 above)
Comment (ekaterina-nbl) [reply lib/Log.js]: In general, they are. But these functions are called from imp-central-api logstream callbacks. In this case, they lose the original context and .bind(this) is really needed.

Comment (ppetrosh) [reply lib/Log.js]: Ah got it, yes. They are called from a different context. Thanks!

File: lib/util/ImpCentralApiHelper.js (hunk 3 above)
Comment (ppetrosh) [reply lib/util/ImpCentralApiHelper.js]: fair enough

File: lib/util/ImpCentralApiHelper.js (hunk 6 above)
Comment (ppetrosh) [reply lib/util/ImpCentralApiHelper.js]: ok sounds reasonable

File: lib/util/ImpCentralApiHelper.js [hunk 7]
```diff
@@ -428,15 +454,24 @@ class ImpCentralApiHelper {
...
         return impCentralApiMethod(...args).
             then(result => this._resolveImpCentralApiResponse(result)).
             catch(error => {
-                if (this._isRateLimitError(error)) {
+                if (this._isAuthenticationError(error)) {
+                    if (refreshAccessTokenOnAuthError) {
+                        // try to refresh access token in case of authentication error and then restart the original request
+                        return this._refreshAccessToken().
+                            then(() => this._processImpCentralApiRequest(
+                                refreshAccessTokenOnAuthError, entityType, id, impCentralApiMethod, ...args));
```
Comment (ppetrosh) [inline lib/util/ImpCentralApiHelper.js]: shouldn't it be `false` instead of `refreshAccessTokenOnAuthError` here (as well as below)?

Comment (ekaterina-nbl) [reply lib/util/ImpCentralApiHelper.js]: I think no. false here means that we will try to refresh expired access token only once per request.
But theoretically access tokens can expire at any moment, eg just after the previous refresh.
Why not to eliminate this case using multiple refresh per request?

Comment (ppetrosh) [reply lib/util/ImpCentralApiHelper.js]: I see ok. You are right.
The only concern that I initially had is that it (in theory) may take forever. Which is probably acceptable at this point.

**Pull request merged**